### PR TITLE
Make `NonUseContext::AscribeUserTy` carry `ty::Variance`

### LINKED ISSUE #108267: `AscribeUserType` statements contain variance which should be carried by `NonUseContext::AscribeUserTy`
_Originally posted by @cjgillot in https://github.com/rust-lang/rust/pull/107969#discussion_r1111769379, see that PR discussion for more detail_

### PR DESCRIPTION
**Pull request by obeis**
Close #108267

### CODE REVIEW

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 1]
```diff
@@ -777,7 +777,7 @@ impl<'a, 'b, 'tcx> TypeVerifier<'a, 'b, 'tcx> {
                 Inspect | Copy | Move | PlaceMention | SharedBorrow | ShallowBorrow | UniqueBorrow
                 | AddressOf | Projection,
             ) => ty::Covariant,
-            PlaceContext::NonUse(AscribeUserTy) => ty::Covariant,
+            PlaceContext::NonUse(AscribeUserTy(_)) => ty::Covariant,
```
Comment (lcnr) [inline compiler/rustc_borrowck/src/type_check/mod.rs:780]: use the variance of the place context here, that's why we want to add the variance in the first place.

File: compiler/rustc_middle/src/mir/visit.rs [hunk 2]
```diff
@@ -149,7 +149,7 @@ macro_rules! make_mir_visitor {
             fn visit_ascribe_user_ty(
                 &mut self,
                 place: & $($mutability)? Place<'tcx>,
-                variance: $(& $mutability)? ty::Variance,
+                variance: ty::Variance,
```
Comment (lcnr) [inline compiler/rustc_middle/src/mir/visit.rs:152]: why remove the `$(& $mutability)?`?

File: compiler/rustc_middle/src/mir/visit.rs [hunk 3]
```diff
@@ -782,12 +782,12 @@ macro_rules! make_mir_visitor {
...
             fn super_ascribe_user_ty(&mut self,
                                      place: & $($mutability)? Place<'tcx>,
-                                     _variance: $(& $mutability)? ty::Variance,
+                                     variance: ty::Variance,
                                      user_ty: & $($mutability)? UserTypeProjection,
                                      location: Location) {
                 self.visit_place(
                     place,
-                    PlaceContext::NonUse(NonUseContext::AscribeUserTy),
+                    PlaceContext::NonUse(NonUseContext::AscribeUserTy(variance)),
```
Comment (lcnr) [inline compiler/rustc_middle/src/mir/visit.rs:790]: you should be able to use `(** &$mutability)? variance` here.

i.e. if there is mutability, add a redundant `&mut` for the macro matcher and then deref that and the original one.

File: compiler/rustc_middle/src/mir/visit.rs [hunk 4]
```diff
@@ -1320,7 +1320,7 @@ pub enum NonUseContext {
     /// Ending a storage live range.
     StorageDead,
     /// User type annotation assertions for NLL.
-    AscribeUserTy,
+    AscribeUserTy(Variance),
```
Comment (cjgillot) [inline compiler/rustc_middle/src/mir/visit.rs:1323]:
```suggestion
    AscribeUserTy(ty::Variance),
```
For uniformity in the file.

**Pull request merged**